dozer-core/src/checkpoint/mod.rs

```diff
@@ -177,68 +169,38 @@ impl CheckpointFactory {
     fn write_record_store_slice(
         &self,
         key: String,
-        source_states: &SourceStates,
+        source_states: SourceStates,
     ) -> Result<(), ExecutionError> {
         let mut state = self.state.lock();
         let (data, num_records_serialized) =
             self.record_store.serialize_slice(state.next_record_index)?;
         state.next_record_index += num_records_serialized;
         drop(state);
 
-        self.write_record_store_slice_data(key, source_states, data)
-            .map_err(|_| ExecutionError::CheckpointWriterThreadPanicked)
-    }
-
-    fn write_record_store_slice_data(
-        &self,
-        key: String,
-        source_states: &SourceStates,
-        data: Vec<u8>,
-    ) -> Result<(), SendError<String>> {
-        let source_states =
-            bincode::serialize(source_states).expect("Source states should be serializable");
-
-        self.queue.create_upload(key.clone())?;
-        self.queue.upload_chunk(
-            key.clone(),
-            (source_states.len() as u64).to_le_bytes().to_vec(),
-        )?;
-        self.queue.upload_chunk(key.clone(), source_states)?;
-        self.queue.upload_chunk(key.clone(), data)?;
-        self.queue.complete_upload(key)?;
+        let data = bincode::serialize(&RecordStoreSlice {
```

Comment:
This will use `serde::Serialize::serialize_seq` for `data`, which will encode elements of `data` one by one instead of just `memcpy`'ing.

You can work around that by either: implementing `Serialize` and `Deserialize` on `RecordStoreSlice`, and use `serialize_bytes` to serialize `data`, or use [serde-bytes](https://docs.rs/serde_bytes/latest/serde_bytes/) with its wrapper types or annotations

Reply:
No wonder I felt slowdown!

Reply:
This optimization does need bincode 2, which is currently in the RC stage

Reply:
Of course you could go the manual route instead of bincode 2 and just copy the len and the buffer into a vec yourself.

Reply:
I'll go for the rc!